.dotnet/src/Custom/Batch/BatchOperation.Protocol.cs

```diff
@@ -0,0 +1,287 @@
+
+        ClientResult result = client.GetBatch(token.BatchId, cancellationToken.ToRequestOptions());
+        PipelineResponse response = result.GetRawResponse();
+
+        using JsonDocument doc = JsonDocument.Parse(response.Content);
+        string status = doc.RootElement.GetProperty("status"u8).GetString()!;
+
+        return new BatchOperation(client.Pipeline, client.Endpoint, token.BatchId, status, response);
+    }
+
+    /// <inheritdoc/>
+    public override async Task WaitForCompletionAsync(CancellationToken cancellationToken = default)
+    {
+        _pollingInterval ??= new();
+
+        while (!IsCompleted)
+        {
+            PipelineResponse response = GetRawResponse();
+
+            await _pollingInterval.WaitAsync(response, cancellationToken);
+
+            ClientResult result = await GetBatchAsync(_batchId, cancellationToken.ToRequestOptions()).ConfigureAwait(false);
+
+            ApplyUpdate(result);
+        }
+    }
+
+    /// <summary>
+    /// Waits for the operation to complete processing on the service.
+    /// </summary>
+    /// <param name="pollingInterval"> The time to wait between sending requests
+    /// for status updates from the service. </param>
+    /// <param name="cancellationToken"> A token that can be used to cancel this
+    /// method call. </param>
+    public async Task WaitForCompletionAsync(TimeSpan pollingInterval, CancellationToken cancellationToken = default)
+    {
+        _pollingInterval = new(pollingInterval);
+
+        await WaitForCompletionAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public override void WaitForCompletion(CancellationToken cancellationToken = default)
+    {
+        _pollingInterval ??= new();
+
+        while (!IsCompleted)
+        {
+            PipelineResponse response = GetRawResponse();
+
+            _pollingInterval.Wait(response, cancellationToken);
+
+            ClientResult result = GetBatch(_batchId, cancellationToken.ToRequestOptions());
+
+            ApplyUpdate(result);
+        }
+    }
+
+    /// <summary>
+    /// Waits for the operation to complete processing on the service.
+    /// </summary>
+    /// <param name="pollingInterval"> The time to wait between sending requests
+    /// for status updates from the service. </param>
+    /// <param name="cancellationToken"> A token that can be used to cancel this
+    /// method call. </param>
+    public void WaitForCompletion(TimeSpan pollingInterval, CancellationToken cancellationToken = default)
+    {
+        _pollingInterval = new(pollingInterval);
+
+        WaitForCompletion(cancellationToken);
+    }
+
+    private void ApplyUpdate(ClientResult result)
+    {
+        PipelineResponse response = result.GetRawResponse();
+
+        using JsonDocument doc = JsonDocument.Parse(response.Content);
+        string? status = doc.RootElement.GetProperty("status"u8).GetString();
+
+        IsCompleted = GetIsCompleted(status);
```

Comment:
Should we disallow calls to other methods on the LRO if the LRO completes? i.e. is it OK to call Cancel (and others) after completion?

Reply:
Some of them are still relevant, for example, I believe things like FineTuningOperation.GetJobCheckpoints and VectorStoreFileBatchOperation.GetFileAssociates.  But others like Cancel, you are right maybe don't make sense.  Would you throw, or would you let the service return whatever response it deems reasonable?  I can see what happens if you try to cancel a completed operation, i.e. does the service return a 4xx code.

Reply:
It looks like for some of these at least, it is valid to cancel after the operation is completed -- in the sense that the service doesn't return a 4xx.

Reply:
For this one, we'll need to investigate operation-by-operation which of these throw if called after the operation completes.   @KrzysztofCwalina makes a good point about it being a bad dev ex if someone has to wrap operations like Cancel in a try-catch to avoid race conditions.

Reply:
Flagging for follow-up separately.